### PR TITLE
Update game state to check if rounds and or match is finished

### PR DESCRIPTION
**Pull request by dbertella**

### CODE REVIEW

File: packages/server/src/rooms.ts [hunk 1]
```diff
@@ -64,3 +64,30 @@ async function doJoinRoom(io: IOServer, socket: Socket, store: Store, room: Room
     io.in(room.name).emit(GameEvent.CurrentState, state);
   }
 }
+
+export function updateGameState(io: IOServer, store: Store, roomName: string, gameState: GameState) {
```
Comment (stathismor) [inline packages/server/src/rooms.ts]: This file is supposed to be about room management, not game state...

File: packages/server/src/Store.ts [hunk 2]
```diff
@@ -41,21 +41,16 @@ export class Store {
...
 
-  updateRoomState(roomName: string, gameState: GameState): GameState | undefined {
+  updateRoomState(roomName: string, gameState: GameState) {
     const room = this.getRoom(roomName);
-
-    if (room.states.length === 0) {
-      return undefined;
-    }
-
-    room.states[room.states.length - 1] = gameState;
```
Comment (stathismor) [inline packages/server/src/Store.ts:58]: :man_facepalming:

File: packages/server/src/rooms.ts [hunk 3]
```diff
@@ -64,3 +64,30 @@ async function doJoinRoom(io: IOServer, socket: Socket, store: Store, room: Room
...
+
+  const isMatchFinished = isRoundFinshed && gameState.table.length === 0;
+
+  if (isMatchFinished) {
+    // TODO send the final score need to figure out who was the last player to capture cards because
+    // all the final cards on the table will need to go to them
+    io.in(roomName).emit(GameStatus.Ended);
+  } else if (isRoundFinshed) {
+    const { deck, players, ...rest } = gameState;
+    const updatedPlayersHands = players.map((player) => ({
```
Comment (stathismor) [inline packages/server/src/rooms.ts]:
```suggestion
    const updatedPlayers = players.map((player) => ({
```
(Cause it's not just the hand)

File: packages/server/src/rooms.ts [hunk 4]
```diff
@@ -64,3 +64,30 @@ async function doJoinRoom(io: IOServer, socket: Socket, store: Store, room: Room
...
+      ...player,
+      hand: deck.splice(0, 3),
+    }));
+    const updatedGameState = {
+      ...rest,
+      players: updatedPlayersHands,
+      deck,
+    };
+    store.updateRoomState(roomName, updatedGameState);
```
Comment (stathismor) [inline packages/server/src/rooms.ts]: Hm... Is the UI going to add some ordered animations, where you first see the opponent playing, and then the new turn happening?

Comment (dbertella) [reply packages/server/src/rooms.ts]: yeah I have to think how to do this, basically we have to add a timeout everytime there is a state update and run some animations before the state is actually applied, quite a pain, not sure how is going to work or what's the best solution

File: packages/server/src/rooms.ts (hunk 1 above)
Comment (dbertella) [reply packages/server/src/rooms.ts]: so should I create a games file? The problem is that game state right now it's coupled with room, and we are creating the game state inside the room already that's why I added it here, maybe I move that part as well

Comment (dbertella) [reply packages/server/src/rooms.ts]: ok that part is in utils actually, so next question is, is there a difference between this updateGameState and generateGameState?

The difference is that this function emit some events to the socket while the other is not. Should we move all the thing that emits to a common folder maybe?

Comment (stathismor) [reply packages/server/src/rooms.ts]: `generateGameState` is kinda temporary. If the game is now playable, it should probably live in a game-related module.

In terms of separation... Yeah, we might need some refactoring. I guess `Game` is anything game logic related, and Room is anything around joining/players etc. I say we merge what we have, and review/refactor later.

File: packages/server/src/rooms.ts (hunk 4 above)
Comment (stathismor) [reply packages/server/src/rooms.ts]: Phaser uses a concept of _tweens_. A tween is a scheduled movement of an object, that can move from point A to B, with various interpolation ways (Linear, Ease, etc), You can have them being delayed, be run in a loop, and also queue them together. So, I think we need a way to queue animations, that reference the same object (for example a card can move from point A to B, and then from where it was to C).

**Review by stathismor — APPROVED** [review]

**Pull request merged**